pkg/asset/manifests/storageclassconfig.go

```diff
@@ -0,0 +1,135 @@
+package manifests
+
+import (
+	"fmt"
+	"path/filepath"
+
+	yaml "github.com/ghodss/yaml"
+	"github.com/pkg/errors"
+
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/installconfig"
+	azuretypes "github.com/openshift/installer/pkg/types/azure"
+)
+
+var (
+	masterStorageClassConfig = filepath.Join(manifestDir, "storageclass-01-master.yaml")
+	workerStorageClassConfig = filepath.Join(manifestDir, "storageclass-02-worker.yaml")
+)
+
+// StorageClass generates the storageclass yaml files.
+type StorageClass struct {
+	FileList []*asset.File
+}
+
+var _ asset.WritableAsset = (*StorageClass)(nil)
+
+// Name returns a human friendly name for the asset.
+func (*StorageClass) Name() string {
+	return "Storage Class"
+}
+
+// Dependencies returns all of the dependencies directly needed to generate the
+// asset.
+func (*StorageClass) Dependencies() []asset.Asset {
+	return []asset.Asset{
+		&installconfig.InstallConfig{},
+		&installconfig.ClusterID{},
+		// PlatformCredsCheck just checks the creds (and asks, if needed)
+		// We do not actually use it in this asset directly, hence
+		// it is put in the dependencies but not fetched in Generate
+		&installconfig.PlatformCredsCheck{},
+	}
+}
+
+// Generate generates the storage class config file.
+func (sc *StorageClass) Generate(dependencies asset.Parents) error {
+	installConfig := &installconfig.InstallConfig{}
+	clusterID := &installconfig.ClusterID{}
+	dependencies.Get(installConfig, clusterID)
+
+	// It's possible to have different storage class types for both the
+	// master pool and the worker pool. For the master pool it is simple to
+	// determine the storage class. For the worker pool, this is a bit of a
+	// hacky solution. As soon as a storage class is determined, we stop
+	// and assume that is the storage class. This should probably be added
+	// as a validation.
+	switch installConfig.Config.Platform.Name() {
+	case azuretypes.Name:
+		masterPool := installConfig.Config.ControlPlane.Platform.Azure
+		if masterPool.OSDisk.DiskEncryptionSet != nil {
+			storageClass := azureStorageClass(masterPool, "master")
+			configData, err := yaml.Marshal(storageClass)
+			if err != nil {
+				return errors.Wrapf(err, "failed to encode %s master manifest", sc.Name())
+			}
+			sc.FileList = append(sc.FileList, &asset.File{
+				Filename: masterStorageClassConfig,
+				Data:     configData,
+			})
+		}
+
+		workerPool := installConfig.Config.Compute
+		for _, pool := range workerPool {
+			wpool := pool.Platform.Azure
+			if wpool.OSDisk.DiskEncryptionSet != nil {
+				storageClass := azureStorageClass(wpool, "worker")
```

Comment:
Perhaps I've taken the wrong approach here. I was mimicking what is done in ARO. There is a storage class for control plane nodes, and another storage class for worker nodes. Ideally a user would use the worker nodes storage class. Let me know your thoughts on how this could be improved.

Reply:
It depend how the encryption works. Can be the volumes from the "worker" class attached to + decrypted by "master" VMs (and the other way around)? Then we may survive just with one SC with the worker key.